internal/api/impl/proxy/proxy.go

```diff
@@ -28,67 +29,187 @@ import (
 	"github.com/perses/perses/internal/api/interface/v1/globaldatasource"
 	"github.com/perses/perses/internal/api/interface/v1/globalsecret"
 	"github.com/perses/perses/internal/api/interface/v1/secret"
+	"github.com/perses/perses/internal/api/shared"
 	"github.com/perses/perses/internal/api/shared/crypto"
 	databaseModel "github.com/perses/perses/internal/api/shared/database/model"
+	"github.com/perses/perses/internal/api/shared/rbac"
 	"github.com/perses/perses/internal/api/shared/route"
 	"github.com/perses/perses/internal/api/shared/utils"
 	v1 "github.com/perses/perses/pkg/model/api/v1"
 	datasourceHTTP "github.com/perses/perses/pkg/model/api/v1/datasource/http"
+	"github.com/perses/perses/pkg/model/api/v1/role"
 	promConfig "github.com/prometheus/common/config"
 	"github.com/sirupsen/logrus"
 )
 
+// unsavedProxyBody is the body of the request when the datasource is not saved yet.
+// It contains the body of the request and the datasource spec, which is used to build the proxy rather than
+// retrieving the datasource from the database.
+type unsavedProxyBody struct {
+	Method string             `json:"method" yaml:"method"`
+	Body   []byte             `json:"body" yaml:"body"`
+	Spec   *v1.DatasourceSpec `json:"spec" yaml:"spec"`
+}
+
+func (u *unsavedProxyBody) setRequestParams(ctx echo.Context) {
+	req := ctx.Request()
+	req.Method = u.Method
+
+	if len(u.Body) > 0 {
+		req.Body = io.NopCloser(strings.NewReader(string(u.Body)))
+	} else {
+		req.Body = nil
+	}
+
+	ctx.SetRequest(req)
+}
+
 type endpoint struct {
 	Dashboard    dashboard.DAO
 	Secret       secret.DAO
 	GlobalSecret globalsecret.DAO
 	DTS          datasource.DAO
 	GlobalDTS    globaldatasource.DAO
 	Crypto       crypto.Crypto
+
```

Comment:
no need for this new line.

and actually we could privatise every attribute of this struct.

```go
type endpoint struct {
	dashboard    dashboard.DAO
	secret       secret.DAO
	globalSecret globalsecret.DAO
	dts          datasource.DAO
	globalDTS    globaldatasource.DAO
	crypto       crypto.Crypto
	rbac rbac.RBAC
}
```

Reply:
Done